pkg/grpc/jwt_authenticator.go

```diff
@@ -0,0 +1,137 @@
+package grpc
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+
+	"github.com/buildbarn/bb-storage/pkg/clock"
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+type JWTKeyConfig struct {
+	Key interface{}
+}
+
+type jwtAuthenticator struct {
+	verifyKeys []JWTKeyConfig
+	clock      clock.Clock
+}
+
+// From: https://github.com/square/go-jose/blob/v2/jose-util/utils.go
+func loadJSONWebKey(json []byte, pub bool) (*jose.JSONWebKey, error) {
+	var jwk jose.JSONWebKey
+	err := jwk.UnmarshalJSON(json)
+	if err != nil {
+		return nil, err
+	}
+	// Commenting out due to issue with .Valid returning false on symmetric keys:
+	// https://github.com/square/go-jose/issues/314
+	// if !jwk.Valid() {
+	// 	return nil, errors.New("invalid JWK key")
+	// }
+	// if jwk.IsPublic() != pub {
+	// 	return nil, errors.New("priv/pub JWK key mismatch")
+	// }
+	return &jwk, nil
+}
+
+// LoadJWTPublicKey loads a public key from PEM/DER/JWK-encoded data.
+// From: https://github.com/square/go-jose/blob/v2/jose-util/utils.go
+func loadJWTPublicKey(data []byte) (interface{}, error) {
+	input := data
+
+	block, _ := pem.Decode(data)
+	if block != nil {
+		input = block.Bytes
+	}
+
+	// Try to load SubjectPublicKeyInfo
+	pub, err0 := x509.ParsePKIXPublicKey(input)
+	if err0 == nil {
+		return pub, nil
+	}
+
+	cert, err1 := x509.ParseCertificate(input)
+	if err1 == nil {
+		return cert.PublicKey, nil
+	}
+
+	jwk, err2 := loadJSONWebKey(data, true)
+	if err2 == nil {
+		return jwk, nil
+	}
+
+	return nil, fmt.Errorf("JWT setup error: parse error, got '%s', '%s' and '%s'", err0, err1, err2)
+}
+
+// NewJWTAuthenticator creates an Authenticator that
+// only grants access in case a validly-signed JWT (JSON Web Token)
+// is passed as a Bearer token in the request's "authorization" header.
+func NewJWTAuthenticator(keys []JWTKeyConfig, clock clock.Clock) Authenticator {
+	return &jwtAuthenticator{
+		verifyKeys: keys,
+		clock:      clock,
+	}
+}
+
+func (a *jwtAuthenticator) Authenticate(ctx context.Context) error {
+	// Get the gRPC metadata.
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return status.Error(codes.Unauthenticated, "Connection was not established using gRPC")
+	}
+
+	// Extract the `authorization` header.
+	// Note: The keys within the metadata are normalized to lowercase.
+	//       https://godoc.org/google.golang.org/grpc/metadata#New
+	authHeader, ok := md["authorization"]
+	if !ok || len(authHeader) < 1 {
+		return status.Error(codes.Unauthenticated, "authorization required")
+	}
+
+	if len(authHeader) > 1 {
+		return status.Error(codes.Unauthenticated, "multiple authorization headers are not supported")
+	}
+
+	if !strings.HasPrefix(authHeader[0], "Bearer ") {
+		return status.Error(codes.Unauthenticated, "authorization required")
```

Comment:
Nit: Please capitalize error messages, for consistency with the rest of the code base. Also make sure these messages are explicit. "No bearer token found in authorization header", for example.

Reply:
Will update for capitalization.

These error messages potentially leak information to attackers about how it handles authentication (although it may not be much information). Where do you "draw the line" between minimizing leaking the operation of the authentication code and errors actually being helpful? I prefer to err on the side of not leaking any information other than the fact that authentication failed.

Reply:
I think it's fine if errors reveal information about the overall structure. Stuff like "No bearer token found in authorization header" doesn't reveal the actual contents of this header. Only that it was malformed in the general sense. There should be no security implications in that.